### PR TITLE
GITHUB-5391: redo association type form

### PR DESCRIPTION
**Pull request by a2xchip**
[//]: <> (<3 Thanks for taking the time to contribute! You're awesome! <3)

[//]: <> (If you've never contributed to this repository before, please read https://github.com/akeneo/pim-community-dev/blob/master/.github/CONTRIBUTING.md) 

**Description (for Contributor and Core Developer)**

See #5391 
- upgrade AssociationType controller
- upgrade AssociationType rest controller
- create association type form extensions
- update requirejs.yml
- implement association type edit form js module and html template
- implement association type save form and saver js modules
- implement association type delete form and remover js modules
- upgrade association type normalizer
- update specs
- update translations
- update behats
- fix translations form event bug
- review fixes

[//]: <> (What does this Pull Request do? reference the related issue?)

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | :white_check_mark:
| Added Behats                      | :white_check_mark:
| Changelog updated                 | :clock1:
| Review and 2 GTM                  | :white_check_mark:
| Micro Demo to the PO (Story only) | :negative_squared_cross_mark:
| Migration script                  | :negative_squared_cross_mark:
| Tech Doc                          | :negative_squared_cross_mark:

:white_check_mark: Done and pass

:red_circle: Done but fail

:clock1: Done but pending

:negative_squared_cross_mark: Not needed

### CODE REVIEW

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/AssociationTypeController.php [hunk 1]
```diff
@@ -44,4 +81,97 @@ public function indexAction()
...
+     *
+     * @AclAncestor("pim_enrich_associationtype_remove")
+     */
+    public function removeAction($code)
+    {
+        $associationType = $this->getAssociationTypeOr404($code);
+
+        $this->remover->remove($associationType);
+
+        return new JsonResponse();
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/AssociationTypeController.php]: `return new JsonResponse(null, 204);`

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/association-type/form/save.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
+'use strict';
+
+/**
+ * Save extension for association type
+ *
+ * @author    Alexandr Jeliuc <alex@jeliuc.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/association-type/form/save.js]: You worked on it last week ! It explains everything ;)

File: src/Pim/Bundle/EnrichBundle/Resources/public/templates/association-type/tab/properties/general.html [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+<div class="tabsection-content">
+    <div class="AknFormContainer AknFormContainer--withPadding">
+        <div class="AknFieldContainer">
+            <div class="AknFieldContainer-header">
+                <label class="control-label required" for="pim_enrich_association_type_form_code">
+                    <%- codeLabel %> <em><%- __('pim_enrich.form.required') %></em>
+                </label>
+            </div>
+            <div class="AknFieldContainer-inputContainer">
+                <input id="pim_enrich_association_type_form_code" class="AknTextField" type="text" readonly="readonly" disabled="disabled" required="required" value="<%- model.code %>">
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/templates/association-type/tab/properties/general.html]: you can replace by `readonly disabled required`

Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/templates/association-type/tab/properties/general.html]: Btw I don't understand why you add disabled

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/association-type/form/save.js (hunk 2 above)
Comment (a2xchip) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/association-type/form/save.js]: :D :D :D

File: src/Pim/Bundle/EnrichBundle/Resources/public/templates/association-type/tab/properties/general.html (hunk 3 above)
Comment (a2xchip) [inline src/Pim/Bundle/EnrichBundle/Resources/public/templates/association-type/tab/properties/general.html]: it is code that is why it is disabled! ;-) 
and i know other place to remove ="*" attribute value :-P

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/AssociationTypeController.php (hunk 1 above)
Comment (a2xchip) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/AssociationTypeController.php]: :+1:

Comment (rybus) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/AssociationTypeController.php]: You should use Response::HTTP_NO_CONTENT constant instead of 204

**Comment by nidup** [comment]
Nice PR, great work @a2xchip! :tada: 

Is the class `Pim\Bundle\EnrichBundle\Form\Type\AssociationTypeType` is still used? if not, you can drop the class, the related service and indicate this BC Break in the changelog.md.

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/form/common/properties/translation.js [hunk 4]
```diff
@@ -78,6 +78,7 @@ define([
                         label: this.config.label,
                         fieldBaseId: this.config.fieldBaseId
                     }));
+                    this.delegateEvents();
```
Comment (a2xchip) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/form/common/properties/translation.js]: @nidup The hardest part of all issue :-P

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js [hunk 5]
```diff
@@ -215,7 +215,7 @@ define(
             },
 
             loadAssociationTypes: function () {
-                return FetcherRegistry.getFetcher('association-type').fetchAll();
+                return FetcherRegistry.getFetcher('association_type').fetchAll();
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js]: you should use association-type

Comment (a2xchip) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js]: Thanks

Comment (a2xchip) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/form/associations.js]: fixed

**Comment by a2xchip** [comment]
@nidup We discussed AssociationTypeType removal with @juliensnz, as a result we think that it is better to get form and form handler back, to close issue, and then create new issue introducing js popin create form and removin all unused form types and form handlers.

**Comment by nidup** [comment]
Ok, no pb

File: features/association-type/display_association_type_history.feature [hunk 6]
```diff
@@ -13,7 +13,7 @@ Feature: Display the association type history
       | Code | REPLACEMENT |
     And I press the "Save" button
```
Comment (a2xchip) [inline features/association-type/display_association_type_history.feature]: @juliensnz Possibly this should be split to have two dedicated scenarios: Successfully edit an association type; Successfully see the history.

**Comment by a2xchip** [comment]
@juliensnz Travis passed \o/

**Comment by nidup** [comment]
congratz @a2xchip ! well done!

**Comment by a2xchip** [comment]
:tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada: :tada:

**Comment by grena** [comment]
Congrats! 🌮

**Comment by anaelChardan** [comment]
Congrats !

**Comment by a2xchip** [comment]
@anaelChardan Thank you!!!!! :-) :-)

**Pull request merged**